### DISCUSSION TITLE
Best way to create the window handle for the apply?

### DISCUSSION THREAD

**Discussion by ds-andreasor** [wixtoolset/issues#8054] [Questions] [answered]
If I want to start the engine apply phase a window handle is needed.
In the v.3 custom BA I used `HwndPtr = new WindowInteropHelper(view).EnsureHandle();` and view was the root wpf window.
This worked fine even when a bundle upgrade was done.

In v.5.0.0-rc.1 this is not good enough anymore.
So now I create a hidden apply window and use the handle of it.
>// Create dummy window to always have a window handle for the apply phase.
ApplyWindow = new Window
{
    Focusable = false,
    ShowInTaskbar = false,
    Title = "BundleApplyWindow",
    Visibility = Visibility.Hidden,
    WindowStyle = WindowStyle.None
};
HwndPtr = new WindowInteropHelper(ApplyWindow).EnsureHandle();

This works for a "normal" install but not for an bundle upgrade.
> [0CE0:1854][2024-03-17T15:52:17]e000: Error 0x80070057: BA passed invalid hwndParent to Apply.
[0CE0:1854][2024-03-17T15:52:17]e000: Error 0x80070057: Failed to detect in the engine.
[0CE0:1854][2024-03-17T15:52:17]e000: Error 0x80070057: BA OnPlanComplete failed.
[0CE0:19BC][2024-03-17T15:52:17]e000: Error 0x80070057: BA OnPlanComplete failed.

So maybe someone has it working in v4 or v5.0.0-rc.1 and can give me a hint what I am doing wrong here.

**Comment by barnson** [comment]
That messages happens when the `IsWindow` API function says it's not a window. WixBA does your first approach: https://github.com/wixtoolset/wix/blob/develop/src/test/burn/WixToolset.WixBA/RootView.xaml.cs

**Reply by ds-andreasor** [reply]
Yes I have checked the code in the WiXBA but I am not sure if the WiXBA is still build and tested for versions >= 4.0.0.
This only happens when a bundle upgrade runs.
The install of bundle A version 1.0.0.0 works with both methods.
It hangs when the upgraded bundle A version 1.0.0.1 installs. From the logs it is the uninstall of the related bundle A version 1.0.0.0 that has the error.

It works for a bundle that uses the WiX StdBA.
I don't see a difference in the detect and plan part of the logs between WiX StdBA bundle and the custom BA bundle.
This is why I am not sure what could be wrong in the custom BA code now.

Maybe the WiXBA has the same issue and it is just not tested.

**Comment by ds-andreasor** [comment]
Here are logs from a test with a WiX StdBA bundle upgrade:
[CAD_CAM_Prerequisites_2.0.1_20240318150633.log](https://github.com/wixtoolset/issues/files/14638217/CAD_CAM_Prerequisites_2.0.1_20240318150633.log)
[CAD_CAM_Prerequisites_2.0.1_20240318150633_002_{6EE583D5-B3E9-4C5D-B245-003F9752762A}.log](https://github.com/wixtoolset/issues/files/14638222/CAD_CAM_Prerequisites_2.0.1_20240318150633_002_.6EE583D5-B3E9-4C5D-B245-003F9752762A.log)

And here from a custom BA bundle (I added some more debug messages to be sure that it is not a custom BA problem):
[inLab_CAM_SW_0.0.1_Build_280185.1_20240318141210.log](https://github.com/wixtoolset/issues/files/14638253/inLab_CAM_SW_0.0.1_Build_280185.1_20240318141210.log)
[inLab_CAM_SW_0.0.1_Build_280185.1_20240318141210_005_{43E41095-9CC2-4D61-9E3B-55D861C1D332}.log](https://github.com/wixtoolset/issues/files/14638258/inLab_CAM_SW_0.0.1_Build_280185.1_20240318141210_005_.43E41095-9CC2-4D61-9E3B-55D861C1D332.log)

I can not spot a difference besides that it fails at the apply.

**Reply by barnson** [reply]
https://github.com/wixtoolset/wix/pull/509 has a PR with a test change to user WixBA during install and upgrade. It works. So the HWND approach it uses is valid.

**Reply by ds-andreasor** [reply]
Thanks Bob for the PR so the code of the toolset is fine and I must have a hidden error in the custom BA code.
I hope that it is a timing issue but still have to run builds and tests to be sure.
Will complete this if I know what is going on.

**Comment by ds-andreasor** [comment, marked as answer]
Okay I finally got it working again.
Just for later reference:
I issue was that the Custom BA was not setting the window handle in quiet/embedded mode that the related bundle needs.
It seems that v3 somehow passed the apply handle to the related bundle and v4 and above does not do this.

**Reply by barnson** [reply]
WiX v3 didn't require a valid window handle for Apply so that makes sense.
